<a href="https://colab.research.google.com/github/jonaleal/ModelosRNA-Kmeans/blob/main/Modelo_RNA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from sklearn.datasets import make_blobs
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.model_selection import GridSearchCV
import matplotlib.pyplot as plt
import numpy as np
from sklearn.metrics import accuracy_score

#MODELO RNA

In [ ]:
X, y = make_blobs(
        n_samples    = 500,
        n_features   = 2,
        centers      = 3,
        cluster_std  = 1.2,
        shuffle      = True,
        random_state = 0
       )

#Divición de datos
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

# Crear un modelo de red neuronal
modelo = MLPClassifier(random_state=0,)
parameters = {'hidden_layer_sizes': [(5,), (10,), (15,), (5, 5), (10, 10), (15, 15), (5, 5, 5), (10, 10, 10), (15, 15, 15)],
              'activation': ['logistic'],
              'solver': ['sgd'],
              'max_iter': [2500]}

# Configurar GridSearchCV
grid_search = GridSearchCV(modelo, parameters, cv=5)

# Entrenar el modelo
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=MLPClassifier(random_state=0),
             param_grid={'activation': ['logistic'],
                         'hidden_layer_sizes': [(5,), (10,), (15,), (5, 5),
                                                (10, 10), (15, 15), (5, 5, 5),
                                                (10, 10, 10), (15, 15, 15)],
                         'max_iter': [2500], 'solver': ['sgd']})

In [ ]:
print(results)

{'mean_fit_time': array([1.638166  , 1.51152959, 1.3629611 , 0.14371276, 0.08123059,
       0.09661193, 0.09830885, 0.20892906, 0.12500563]), 'std_fit_time': array([0.76431635, 0.28914955, 0.46031748, 0.08227066, 0.00672677,
       0.01083943, 0.00590349, 0.0115749 , 0.00573086]), 'mean_score_time': array([0.00125413, 0.00122309, 0.00111203, 0.00152874, 0.00102553,
       0.00135298, 0.00202212, 0.00131226, 0.00134859]), 'std_score_time': array([4.19037549e-04, 3.38063589e-04, 2.34247068e-04, 8.83633267e-04,
       1.45954674e-04, 2.61411785e-04, 7.17304579e-04, 4.24360960e-05,
       3.86845735e-05]), 'param_activation': masked_array(data=['logistic', 'logistic', 'logistic', 'logistic',
                   'logistic', 'logistic', 'logistic', 'logistic',
                   'logistic'],
             mask=[False, False, False, False, False, False, False, False,
                   False],
       fill_value='?',
            dtype=object), 'param_hidden_layer_sizes': masked_array(data=[(5,),

In [ ]:
# Obtener los resultados de la búsqueda de GridSearch
results = grid_search.cv_results_
params = results['params']
mean_test_score = results['mean_test_score']
std_test_score = results['std_test_score']

#Extraer la cantidad de neuronas de las configuraciones
neuron_counts = [param['hidden_layer_sizes'] if isinstance(param['hidden_layer_sizes'], tuple) else (param['hidden_layer_sizes'],) for param in params]

#Imprimir la precisión promedio para cada configuración de neuronas
print("Precisión promedio para cada configuración de neuronas:")
for neurons, score, std in zip(neuron_counts, mean_test_score, std_test_score):
    print(f"Neuronas: {neurons}, Precisión: {score:.4f}, Desviación: {std:.4f}")


Precisión promedio para cada configuración de neuronas:
Neuronas: (5,), Precisión: 0.8600, Desviación: 0.0291
Neuronas: (10,), Precisión: 0.8743, Desviación: 0.0246
Neuronas: (15,), Precisión: 0.8771, Desviación: 0.0294
Neuronas: (5, 5), Precisión: 0.2171, Desviación: 0.0398
Neuronas: (10, 10), Precisión: 0.1743, Desviación: 0.0672
Neuronas: (15, 15), Precisión: 0.3457, Desviación: 0.0057
Neuronas: (5, 5, 5), Precisión: 0.3657, Desviación: 0.0508
Neuronas: (10, 10, 10), Precisión: 0.3457, Desviación: 0.0057
Neuronas: (15, 15, 15), Precisión: 0.3457, Desviación: 0.0057


In [ ]:
print("Mejor combinación de hiperparámetros:", grid_search.best_params_)
print("Mejor puntuación de validación cruzada:", grid_search.best_score_)

Mejor combinación de hiperparámetros: {'activation': 'logistic', 'hidden_layer_sizes': (15,), 'max_iter': 2500, 'solver': 'sgd'}
Mejor puntuación de validación cruzada: 0.8771428571428572


Aunque el mejor Modelo para Grid Search sea el que tiene 1 capa y 15 neuronas, Para nosotros el mejor modelo sería el de 1 capa y 10 neuronas puesto que es más simple pero entrega un resultado con muy poca diferentecia de precision tan solo 0.003.

In [ ]:
#Volver a Entrenar con el modelo elegido
hyperparametros = {'activation': 'logistic', 'hidden_layer_sizes': (10,), 'max_iter': 2500, 'solver': 'sgd'}

# Crear un nuevo modelo con los hiperparámetros especificados
elegido = MLPClassifier(**hyperparametros)

# Entrenar el modelo con tus datos de entrenamiento
elegido.fit(X_train, y_train)

# Hacer predicciones con el modelo
prediccion_elegido = elegido.predict(X_test)

# Calcular la precisión del modelo
presicion = accuracy_score(y_test, prediccion_elegido)
print(f"Precisión del modelo específico: {presicion:.4f}")

Precisión del modelo específico: 0.8667


Podemos ver que la Precisión del modelo a la hora de testear es 0.8667 mientras que en el entrenamiento fue 0.874, la diferencia es muy poco y en este caso es menor a la de entrenamiento lo que significa que el modelo no esta sobreajustado